frontend/readhistory.lua

```diff
@@ -178,129 +179,131 @@ function ReadHistory:getLastFile()
-             return self.hist[i].file
+             return v.file
         end
     end
 end
 
+function ReadHistory:updateItemByPath(old_path, new_path)
+    local index = self:getIndexByFile(old_path)
+    if index then
+        self.hist[index].file = new_path
+        self.hist[index].text = new_path:gsub(".*/", "")
+        self.hist[index].callback = function()
+            selectCallback(new_path)
+        end
+        self:_flush()
+        self:reload(true)
+    end
+    if G_reader_settings:readSetting("lastfile") == old_path then
+        G_reader_settings:saveSetting("lastfile", new_path)
+    end
+    self:ensureLastFile()
+end
+
+--- Updates the history list after deleting a file.
 function ReadHistory:fileDeleted(path)
-    if G_reader_settings:isTrue("autoremove_deleted_items_from_history") then
-        self:removeItemByPath(path)
-    else
-        -- Make it dimed
-        for i=1, #self.hist do
-            if self.hist[i].file == path then
-                self.hist[i].dim = true
-                break
-            end
+    local index = self:getIndexByFile(path)
+    if index then
+        if G_reader_settings:isTrue("autoremove_deleted_items_from_history") then
+            self:removeItem(self.hist[index], index)
+        else
+            self.hist[index].dim = true
+            self:ensureLastFile()
         end
-        self:ensureLastFile()
     end
 end
 
+--- Removes the history item if the document settings has been reset.
 function ReadHistory:fileSettingsPurged(path)
     if G_reader_settings:isTrue("autoremove_deleted_items_from_history") then
         -- Also remove it from history on purge when that setting is enabled
         self:removeItemByPath(path)
     end
 end
 
+--- Checks the history list for deleted files and removes history items respectively.
 function ReadHistory:clearMissing()
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == nil or lfs.attributes(self.hist[i].file, "mode") ~= "file" then
-            self:removeItem(self.hist[i], i)
+    local history_updated
+    for i, v in ipairs(self.hist) do
+        if v.file == nil or lfs.attributes(v.file, "mode") ~= "file" then
+            self:removeItem(v, i, true) -- no flush
+            history_updated = true
         end
     end
-    self:ensureLastFile()
-end
-
-function ReadHistory:removeItemByPath(path)
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == path then
-            self:removeItem(self.hist[i])
-            break
-        end
+    if history_updated then
+        self:_flush()
+        self:ensureLastFile()
     end
-    self:ensureLastFile()
 end
 
-function ReadHistory:updateItemByPath(old_path, new_path)
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == old_path then
-            self.hist[i].file = new_path
-            self.hist[i].text = new_path:gsub(".*/", "")
-            self:_flush()
-            self:reload(true)
```

Comment:
Any reason for you to have moved this function up in the file (which make it harder to grasp the changes) ? (Asking in case this was not voluntarily.)
(Unless it is just github/git diff that has a hard time finding what has changed.)

Reply:
To place funtions more logically: `removeItem` near `removeItemByPath`.
(I'm still struggling with spec, please do not review intermediate commits)